src/h3lib/lib/localij.c

```diff
@@ -0,0 +1,551 @@
+    {0, 0, 0, 0, 0, 0, 0},         // 6
+};
+/**
+ * Reverse base cell direction -> leading index digit -> roations 60 ccw.
+ * For reversing the rotation introduced in PENTAGON_ROTATIONS when the index is
+ * on a pentagon.
+ */
+const int PENTAGON_ROTATIONS_REVERSE_NONPOLAR[7][7] = {
+    {0, 0, 0, 0, 0, 0, 0},         // 0
+    {-1, -1, -1, -1, -1, -1, -1},  // 1
+    {0, 1, 0, 0, 0, 0, 0},         // 2
+    {0, 1, 0, 0, 0, 1, 0},         // 3
+    {0, 5, 0, 0, 0, 0, 0},         // 4
+    {0, 1, 0, 5, 1, 1, 0},         // 5
+    {0, 0, 0, 0, 0, 0, 0},         // 6
+};
+/**
+ * Reverse base cell direction -> leading index digit -> roations 60 ccw.
+ * For reversing the rotation introduced in PENTAGON_ROTATIONS when the index is
+ * on a polar pentagon.
+ */
+const int PENTAGON_ROTATIONS_REVERSE_POLAR[7][7] = {
+    {0, 0, 0, 0, 0, 0, 0},         // 0
+    {-1, -1, -1, -1, -1, -1, -1},  // 1
+    {0, 1, 1, 1, 1, 1, 1},         // 2
+    {0, 1, 0, 0, 0, 1, 0},         // 3
+    {0, 1, 0, 0, 1, 1, 1},         // 4
+    {0, 1, 0, 5, 1, 1, 0},         // 5
+    {0, 1, 1, 0, 1, 1, 1},         // 6
+};
+
+// Simply prohibit many pentagon distortion cases rather than handling them.
+const bool FAILED_DIRECTIONS_II[7][7] = {
+    {false, false, false, false, false, false, false},  // 0
+    {false, false, false, false, false, false, false},  // 1
+    {false, false, false, false, true, false, false},   // 2
+    {false, false, false, false, false, false, true},   // 3
+    {false, false, false, true, false, false, false},   // 4
+    {false, false, true, false, false, false, false},   // 5
+    {false, false, false, false, false, true, false},   // 6
+};
+const bool FAILED_DIRECTIONS_III[7][7] = {
+    {false, false, false, false, false, false, false},  // 0
+    {false, false, false, false, false, false, false},  // 1
+    {false, false, false, false, false, true, false},   // 2
+    {false, false, false, false, true, false, false},   // 3
+    {false, false, true, false, false, false, false},   // 4
+    {false, false, false, false, false, false, true},   // 5
+    {false, false, false, true, false, false, false},   // 6
+};
+
+/**
+ * Produces ijk+ coordinates for an index anchored by an origin.
+ *
+ * The coordinate space used by this function may have deleted
+ * regions or warping due to pentagonal distortion.
+ *
+ * Coordinates are only comparable if they come from the same
+ * origin index.
+ *
+ * Failure may occur if the index is too far away from the origin
+ * or if the index is on the other side of a pentagon.
+ *
+ * @param origin An anchoring index for the ijk+ coordinate system.
+ * @param index Index to find the coordinates of
+ * @param out ijk+ coordinates of the index will be placed here on success
+ * @return 0 on success, or another value on failure.
+ */
+int h3ToLocalIjk(H3Index origin, H3Index h3, CoordIJK* out) {
+    if (H3_GET_MODE(origin) != H3_HEXAGON_MODE ||
+        H3_GET_MODE(h3) != H3_HEXAGON_MODE) {
```

Comment:
If we have the ability to validate some user inputs, is there a reason we would not want to? I think some functions aren't able to cleanly report failures right now, which certainly limits where we can add checks (until 4.0.0 and #104)

Reply:
Yes, actually. If the function could be conceivably used in a looping hotspot I could see users want to have a faster iteration time if they're confident in the data already being valid.

This feels like one of those functions. H3 indexes could already be validated and stored in DB, then they want fast conversion to IJK for some planar operations and then back again.

Reply:
I should have replied directly to the performance concern rather than asking a question. I think in this specific case I am OK with removing the mode check because it doesn't appear this will cause issues, for example there are no array accesses based on the mode of the indexes.

Aside: This code was not introduced in this PR, it was introduced in #83 where we had a similar discussion but didn't come to a conclusion as @nrabinowitz suggested.

I believe there are at least some inputs we should validate, a good example being where there are array accesses based on the inputs, since that could potentially result in segfault, etc.